lib/ecto/parameterized_type.ex

```diff
@@ -0,0 +1,82 @@
+defmodule Ecto.ParameterizedType do
+  @moduledoc """
+  Parameterized types are similar to Ecto Custom Types, however they allow a set of options to be
+  specified in the schema which are initialized on compilation and passed to the callback functions
+  as a final argument. Note that options should be specified as a keyword, but it is idiomatic to
+  convert them to maps inside init for easier pattern matching in other callbacks.
+
+  Note that as opposed to regular types, custom types are responsible for handling nil in both load
+  and dump callbacks.
+
+  ## Examples
+
+    To create a parameterized type, create a module as shown below:
+
+    defmodule MyApp.MyType do
+      use Ecto.ParameterizedType
+
+      def type(_params), do: :string
+
+      def init(opts) do
+        validate_opts(opts)
+        Enum.into(opts, %{})
+      end
+
+      def cast(data, params) do
+        ...
+        cast_data
+      end
+
+      def load(data, _loader, params) do
+        ...
+        {:ok, loaded_data}
+      end
+
+      def dump(data, dumper, params) do
+        ...
+        {:ok, dumped_data}
+      end
+
+      def equal?(a, b, _params) do
+        a == b
+      end
+    end
+
+  To use this type in a schema field, specify the type and parameters like this:
+
+    schema "foo" do
+      field "bar", MyApp.MyType, opt1: :baz, opt2: :boo
+    end
+
+  """
+
+  @type opts :: keyword()
+
+  @type params :: term()
+
+  @callback init(opts :: opts()) :: params()
+
+  @callback cast(data :: term, params :: params()) ::
+              {:ok, term} | {:error, keyword()} | :error
+
+  @callback load(value :: any(), loader :: function(), params :: params()) :: {:ok, value :: any()} | :error
+
+  @callback dump(value :: any(), dumper :: function(), params :: params()) :: {:ok, value :: any()} | :error
+
+  @callback type(params :: params()) :: Ecto.Type.t()
+
+  @callback equal?(value1 :: any(), value2 :: any(), params :: params()) :: boolean()
+
+  @callback embed_as(format :: atom(), params :: params()) :: :self | :dump
+
+  @doc false
+  defmacro __using__(_) do
+    quote location: :keep do
+      @behaviour Ecto.ParameterizedType
+      def embed_as(_, _), do: :self
+      def equal?(term1, term2, _params), do: term1 == term2
```

Comment:
I have a hunch we won't need `equal?/3` once we add `cast/3` and `change/3` because those callbacks can simply return `ignore` instead. So can you please add a TODO?

```suggestion
      # TODO: evaluate if we should keep this once we add cast/3 and change/3
      def equal?(term1, term2, _params), do: term1 == term2
```